### ISSUE TITLE
[LibOS] fork() returns positive pid when child process fails to receive checkpoint

### ISSUE THREAD

**Issue by llly** [oscarlab/graphene#2078]
## Description of the problem
When very big enclave calls fork, sometimes it fail during sending/receiving checkpoint.

Using commit several months ago, the thread which calls fork() prints log `failed sending checkpoint (ret = -22)` and `return from shim_vfork(...) = -22`. Child process prints log `failed restoring checkpoint at vma (-22)`.

Now  child process prints log `failed restoring checkpoint at vma (-22)` but parent doesn't print any error log, fork() also return positive pid.

The case is 32G enclave running Java. Java uses Processbuilder(), it calls fork() then execve(). Here is a snap of log by `egrep "vfork|execve|checkpoint"`
```
[P2875:T2:java] ---- start time: 9481 ms shim_vfork() ... vfork
[P2875:T2:java] vfork was called by the application, implemented as an alias to fork in Graphene
[P2875:T2:java] allocating checkpoint store (size = 67108864, reserve = 33554432)
[P2875:T2:java] complete checkpointing data
[P2875:T2:java] checkpoint of 661152 bytes created
[P2931:java] checkpoint mapped at 0xfdffa0000-0xfe00416a0
[P2931:java] read checkpoint of 661152 bytes from parent
[P2931:java] restored memory from checkpoint
[P2931:java] restoring checkpoint at 0xfdffa0000 rebased from 0xfdffa0000
[P2931:T37:java] successfully restored checkpoint at 0xfdffa0000 - 0xfe00416a0
[P2875:T2:java] ---- end time: 113284 ms return from shim_vfork(...) = 37
[P2931:T37:java] ---- start time: 0 ms shim_execve("//setsid",[setsid,bash,-c,echo $$,],[LD_LIBRARY_PATH=/lib:/lib/x86_64-linux-gnu:/usr//lib/x86_64-linux-gnu:/opt/jdk8:/opt/jdk8/lib/amd64/jli,PATH=/:/opt/jdk8/bin:/usr/sbin:/usr/bin:/sbin:/bin,]) ... execv[P2931:T37:java] e
[P2931:T37:java] execve() in a new process
[P2931:T37:java] allocating checkpoint store (size = 67108864, reserve = 33554432)
[P2931:T37:java] complete checkpointing data
[P2931:T37:java] checkpoint of 27432 bytes created
[P2933:setsid] checkpoint mapped at 0xb26f000-0xb275b28
[P2933:setsid] read checkpoint of 27432 bytes from parent
[P2933:setsid] restored memory from checkpoint
[P2933:setsid] restoring checkpoint at 0x0b26f000 rebased from 0xfe03df000
[P2931:T37:setsid] successfully restored checkpoint at 0x0b26f000 - 0x0b275b28
[P2931:T37:java] Temporary process 2931 is exiting after emulating execve (by forking new process to replace this one); will wait for forked process to exit...
[P2875:T73:java] ---- start time: 391583 ms shim_vfork() ... vfork
[P2875:T73:java] vfork was called by the application, implemented as an alias to fork in Graphene
[P2875:T73:java] allocating checkpoint store (size = 67108864, reserve = 33554432)
[P2875:T73:java] checkpoint epoll: 0xfe5ff2758 -> 0xfdff21268 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff2a58 -> 0xfdff21bc8 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff2158 -> 0xfdff222f0 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff3058 -> 0xfdff22a18 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff34d8 -> 0xfdff23140 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff3958 -> 0xfdff23868 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff3dd8 -> 0xfdff23f90 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff4258 -> 0xfdff246b8 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff46d8 -> 0xfdff24de0 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5ff4b58 -> 0xfdff25740 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f290d8 -> 0xfdff25e68 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f29558 -> 0xfdff26590 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f299d8 -> 0xfdff26cb8 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f29e58 -> 0xfdff273e0 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f2a2d8 -> 0xfdff27b08 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f2a758 -> 0xfdff28230 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f2bad8 -> 0xfdff29b20 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f2bf58 -> 0xfdff2a248 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f2c3d8 -> 0xfdff2a970 (base = 0xfdfe54000)
[P2875:T73:java] checkpoint epoll: 0xfe5f2c858 -> 0xfdff2b098 (base = 0xfdfe54000)
[P2875:T73:java] complete checkpointing data
[P2875:T73:java] checkpoint of 958080 bytes created
[P3013:java] checkpoint mapped at 0xfdfe54000-0xfdff3de80
[P3013:java] read checkpoint of 958080 bytes from parent
[P3013:java] restored memory from checkpoint
[P3013:java] restoring checkpoint at 0xfdfe54000 rebased from 0xfdfe54000
[P3013:java] failed restoring checkpoint at vma (-22)
[P3013:java] Error during shim_init() in receive_checkpoint_and_restore (-22)
[P2875:T73:java] ---- end time: 519935 ms return from shim_vfork(...) = 74
```
First successful fork from `[P2875:T2:java]` created `[P2931:java]`, whose pid is 37.
Second failed fork from `[P2875:T73:java]` created `[P3013:java]`. It failed to receive checkpoint, but parent still returns 74.

I also added more log to catch this error. It come from `read_exact` in shim_checkpoint.c. `read` over socket returns EINVAL.

I also added time stamp to track fork time. I can send you the full log if necessary.

**Comment by dimakuv** [comment]
Hm, this is surprising. I see that the error checking is correct: https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/shim_checkpoint.c#L652-L656

**Comment by dimakuv** [comment]
Ah, wait, the current behavior seems correct.

The parent process was able to correctly create the child process and send the checkpoint to the child. So from the perspective of the parent everything is fine. The parent did its job.

The child process received the checkpoint correctly. The error happens during `restoring the checkpoint`. This is the problem of the child, it has nothing to do with the parent. Something is wrong inside the child process.

So I'd say that the Graphene behavior of `several months ago` was wrong.

**Comment by dimakuv** [comment]
Btw, @llly , could you check the latest master from today? We have just landed a Very Big Change to Graphene where there is only a single manifest for the whole application now. (See https://github.com/oscarlab/graphene/pull/2050.)

This is a breaking change, so you'll need to change your manifest file a bit. But it's much easier now to port applications like Java inside Graphene. This change may also fix the problem with the child process you describe here.

**Comment by llly** [comment]
@dimakuv In the whole log,  there is an item shows the parent received wrong pid of child, should be 3013, but now 0.  Parent can catch this failure.
```
[P2875:T2:java] Adding port (handle 0xfe6c65050) for process 2931 (type 0006)
[P2875:T2:java] ---- end time: 113284 ms return from shim_vfork(...) = 37
.........
[P3013:java] failed restoring checkpoint at vma (-22)
[P3013:java] Error during shim_init() in receive_checkpoint_and_restore (-22)
........
[P2875:T73:java] Adding port (handle 0xfe6c7f450) for process 0 (type 0006)
[P2875:T73:java] ---- end time: 519935 ms return from shim_vfork(...) = 74
```

And I'll try latest master.

**Comment by dimakuv** [comment]
> The log item with @@@ is added by myself. I think parent want to read 4 bytes pid from child socket, but only received pal_errno PAL_ERROR_ENDOFSTREAM because child process didn't work. However error code translation returns success error code.

Ah, yes, this is a bug. I think the easy way to fix this is to change this https://github.com/oscarlab/graphene/blob/master/LibOS/shim/src/shim_checkpoint.c#L221-L227:
```
static int read_exact(PAL_HANDLE handle, void* buf, size_t size) {
    size_t bytes = 0;

    while (bytes < size) {
        PAL_NUM x = DkStreamRead(handle, 0, size - bytes, (char*)buf + bytes, NULL, 0);
        if (x == PAL_STREAM_ERROR) {
            /* ============== add something like below ============== */
            if (PAL_NATIVE_ERRNO() == PAL_ERROR_ENDOFSTREAM) {
                return -EPERM;
            }
            /* ============== [end of changes] ============== */
            int err = PAL_ERRNO();
            if (err == EINTR || err == EAGAIN || err == EWOULDBLOCK) {
                continue;
            }
            return -err;
        }

        bytes += x;
    }

    return 0;
}
```

**Comment by dimakuv** [comment]
@llly Are you working on a PR to fix this (and #2131)?
